### DISCUSSION TITLE
Zsh and Carapace: What should I turn off/remove?

### DISCUSSION THREAD

**Discussion by mathomp4** [carapace-sh/carapace-bin#2571] [General]
All,

I recently saw carapace show up in the new formulae for brew and thought, oooh, looks cool!

But, my zsh setup is sort of a mish-mash of all sorts of things including zsh-completions, zsh-autosuggestions, fzf-tab, etc.

So I was wondering, is there any zsh plugin, etc. I should look out for that conflict with carapace? I don't want to try out carapace and draw the wrong opinion because I had other plugins causing issues!

**Comment by rsteube** [comment]
Carapace integrates into the standard zsh completion framework, so for the most part things should just work.

## Tagging and colored highlighting

Carapace makes extensive use of tagging and colored highlighting.
Setting the `:completion:*` zstyle is recommended for the group titles.

[fzf-tab](https://github.com/Aloxaf/fzf-tab) has it's own mechanism to highlight groups so colors might differ.
Colored highlighting might also have some issues ([Aloxaf/fzf-tab#32](https://github.com/Aloxaf/fzf-tab/issues/32)) as I think the processing was only triggered for _"files"_.

## Custom file completion

Carapace has its own file completion and in the sense of zsh these are just values.
So some plugins related to file completion might not trigger (see colored highlighting above).

## Central registry

Carapace acts like a registry. So completions are primarily registered in carapace and then subsequently by it in Zsh.
This enables invocation of completion for embedding (e.g. `sudo <command> <completion of command>...`).
Doing this (well) using the shell is a [near impossible task](https://github.com/carapace-sh/carapace-bridge).

Normal completions registered directly in zsh still work, but as carapace overrides `sudo` completion these would be missing there.
Workaround for this is using the implicit bridge (`export CARAPACE_BRIDGES='zsh'`) but that has some limitations.

## Order of registration

Order of sourcing plugins can be an issue (at least in Powershell).

For completions the last one to register overrides previous ones.
`fpath` should have a lower precedence so I don't expect `zsh-completions` poses an issue.


So if you want to use the default completion of a command that carapace provides you either have to:
- source it after carapace
- disable it in carapace with (`export CARAPACE_EXCLUDES='command'`)

## Completers

`git` and `gh` currently get most of my attention so playing with those would be best to see what is possible with carapace.
Others might be slightly out of date.

Comparing the completion with [elvish](https://elv.sh/) might be useful to see if something is amiss (that one works best and does not need plugins).

> btw. i botched the side-by-side view of flags for zsh in a recent release - this is fixed in master/next version

**Reply by mathomp4** [reply]
Wow. Thanks for the reply! As I test, do you recommend, say, disabling fzf-tab? I'm not really wed to much in my zsh setup as it really is a mixture of "ooh cool feature I saw". 

And I've never heard of elvish, I'll give it a shot as well to see the full power of carapace!

**Reply by rsteube** [reply]
`fzf-tab` should work alright, it's mostly the group and colored highlighting that is impacted.
